### PR TITLE
[language][VM] Split borrow global into Imm and Mut versions

### PR DESCRIPTION
**Pull request by tnowacki**
## Motivation

- Split BorrowGlobal into MutBorrowGlobal and ImmBorrowGlobal
- In the IR, `borrow_global` is now the immutable version, and `borrow_global_mut` is the mutable version.
  - I feel this matches our other syntax around borrowing, i.e. `&` versus `&mut `
- This split will help with immutable reads under the new `acquires` restrictions

## Test Plan

- New tests
- cargo check; cargo test

## Related PRs

Follow up to 
#672 
#557

### CODE REVIEW

**Comment by vgao1996** [comment]
I had a look at the following parts and they seem correct to me! (Thankfully since we're splitting an existing instruction into two similar ones, we can reuse a lot of the old code and make a copy for each version at many places.)
- bytecode_verifier
- compiler & syntax
- stdlib modules
- tools/cost_synthesis
- vm
- vm_runtime

I'll examine the tests tomorrow. At meanwhile, it would be very helpful if you could fix formatting so we can get some signals on tests.

You'll need resolve those conflicts due to a PR I just landed. I had a look and it should be quite trivial. (Just copy pasting & filling the holes.)

**Comment by vgao1996** [comment]
Just as a reminder, since this is a breaking change, you'll need to update the script hashes & the genesis blob before landing.

File: language/bytecode_verifier/src/type_memory_safety.rs [hunk 1]
```diff
@@ -781,30 +781,64 @@ impl<'a> TypeAndMemorySafetyAnalysis<'a> {
...
+            }
+            Bytecode::ImmBorrowGlobal(idx, _) => {
+                let struct_definition = self.module().struct_def_at(*idx);
+                if !StructDefinitionView::new(self.module(), struct_definition)
+                    .is_nominal_resource()
+                {
+                    return Err(VMStaticViolation::BorrowGlobalNoResourceError(offset));
                 }
+
+                let freeze_ok = state
```
Comment (vgao1996) [inline language/bytecode_verifier/src/type_memory_safety.rs]: Not blocking but I did find the name of `fn freeze_ok` a bit confusing. The function checks if there're any mutable references stored in variables or on stack. Should we change it to something more straightforward, like `fn exists_mutable_references`?

Comment (tnowacki) [reply language/bytecode_verifier/src/type_memory_safety.rs]: There needs to be some collapsing of names here, but I prefer succinct judgments that we utilize and compose throughout the rules. In other words `exists_mutable_references` is too imperative for me. 

I would prefer us to use terms like `readable`, `writable`, and `freezable` and then compose them.

**Comment by tnowacki** [comment]
@vgao1996, do you mean `ImmBorrowGlobal`?

**Comment by vgao1996** [comment]
@tnowacki Yes. Sorry for the typo.

**Comment by tnowacki** [comment]
I updated the test to actually run

File: language/bytecode_verifier/src/type_memory_safety.rs [hunk 2]
```diff
@@ -829,13 +829,15 @@ impl<'a> TypeAndMemorySafetyAnalysis<'a> {
                 }
             }
 
-            Bytecode::BorrowGlobal(idx, type_actuals_idx) => {
+            // TODO: Handle type actuals for generics
```
Comment (vgao1996) [inline language/bytecode_verifier/src/type_memory_safety.rs]: Could you delete this comment? This now supports generics.

Comment (runtian-zhou) [reply language/bytecode_verifier/src/type_memory_safety.rs]: > LGTM. One thing I'm unsure. @runtian-zhou is it still required to copy paste the script hashes?

Yea I think so. Though I'm not sure if this diff will change script hashes tho as the simple transaction scripts won't use those bytecode primitives.

File: language/bytecode_verifier/src/type_memory_safety.rs [hunk 3]
```diff
@@ -845,17 +846,52 @@ impl<'a> TypeAndMemorySafetyAnalysis<'a> {
...
+                let nonce = self.get_nonce(&mut state);
+                state.borrow_from_global_value(*idx, nonce.clone());
+                self.stack.push(StackAbstractValue {
+                    signature: SignatureToken::MutableReference(Box::new(struct_type)),
+                    value: AbstractValue::Reference(nonce),
+                });
+                Ok(())
+            }
+
+            Bytecode::ImmBorrowGlobal(idx, type_actuals_idx) => {
```
Comment (shazqadeer) [inline language/bytecode_verifier/src/type_memory_safety.rs]: I wonder if it is possible to factor out common checking code between MutBorrowGlobal and ImmBorrowGlobal.  Something like this was done for MutBorrowLoc and ImmBorrowLoc.

File: language/compiler/ir_to_bytecode/syntax/src/ast.rs [hunk 4]
```diff
@@ -289,9 +289,9 @@ pub enum Builtin {
     /// Check if there is a struct object (`StructName` resolved by current module) associated with
     /// the given address
     Exists(StructName, Vec<Type>),
-    /// Get the struct object (`StructName` resolved by current module) associated with the given
-    /// address
-    BorrowGlobal(StructName, Vec<Type>),
+    /// Get a reference to the resource(`StructName` resolved by current module) associated
+    /// with the given address
+    BorrowGlobal(bool, StructName, Vec<Type>),
```
Comment (shazqadeer) [inline language/compiler/ir_to_bytecode/syntax/src/ast.rs]: Expand the comment to indicate what the bool is for?

Comment (tnowacki) [reply language/compiler/ir_to_bytecode/syntax/src/ast.rs]: Will do

File: language/bytecode_verifier/src/type_memory_safety.rs (hunk 3 above)
Comment (tnowacki) [reply language/bytecode_verifier/src/type_memory_safety.rs]: I good idea, I was a bit too hasty in not doing so

**Comment by vgao1996** [comment]
@bors-libra delegate+

Trying this out. @tnowacki see if you can make trivial changes and land without asking us to approve it again

File: language/bytecode_verifier/src/type_memory_safety.rs [hunk 5]
```diff
@@ -149,12 +149,14 @@ impl<'a> TypeAndMemorySafetyAnalysis<'a> {
     }
 
     // helper for both `ImmBorrowFIeld` and `MutBorrowField`
```
Comment (shazqadeer) [inline language/bytecode_verifier/src/type_memory_safety.rs]: ImmBorrowFIeld -> ImmBorrowField

**Pull request closed without merging**